### PR TITLE
fix(developer): in model compiler, give correct key to shorter prefix words when a longer, higher-frequency word is also present

### LINKED ISSUE #11073: bug(developer): prefix words will be corrupted by the model compiler when a longer, higher frequency word is also present
When a wordlist entry has lower-frequency than a different word that it is a prefix for, our current lexical-model compilation does not properly include the former entry within the compiled model.

This is visible within https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/common/models/templates/test/fixtures/tries/english-1000.json.  Unfortunately, pretty-printing is pretty necessary to actually see the issue.

A pretty-print of the third such entry within that test fixture:

```javascript
"r": {
  "type": "internal",
  "weight": 634,
  "values": [
    "d",
    "t",
    null
  ],
  "children": {
    "d": {
      "type": "leaf",
      "weight": 634,
      "entries": [
        {
          "key": "heard",
          "weight": 634,
          "content": "heard"
        }
      ]
    },
    "t": {
      "type": "leaf",
      "weight": 453,
      "entries": [
        {
          "key": "heart",
          "weight": 453,
          "content": "heart"
        }
      ]
    },
    "undefined": {
      "type": "leaf",
      "weight": 369,
      "entries": [
        {
          "key": "hear",
          "weight": 369,
          "content": "hear"
        }
      ]
    }
  }
},
```

Note that "hear" is frequency 369, while "heart" is frequency 453.

1. Within `"values"`, the `null` should instead be `"\ufdd0"`:
    https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/common/models/templates/src/trie-model.ts#L331-L335
    https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/common/models/templates/src/common.ts#L6
    https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/developer/src/kmc-model/src/build-trie.ts#L274-L284
2. The `"undefined"` child should instead be keyed as `"\ufdd0"`, matching its corresponding key within `"values"`.

As the "values" key does not match the child, such entries are not found by the live predictive-text engine.  This can easily be seen with our English MTNT model by simply typing the word "thin":
- https://github.com/keymanapp/lexical-models/blob/cf2cbb5cdd1fb5cc65cf6d41b84fb2542e8f8ec0/release/nrc/nrc.en.mtnt/source/mtnt.tsv#L55
    - `think: 8872`
- https://github.com/keymanapp/lexical-models/blob/cf2cbb5cdd1fb5cc65cf6d41b84fb2542e8f8ec0/release/nrc/nrc.en.mtnt/source/mtnt.tsv#L2248
    - `thin: 122`
- The word `thin` will not appear as an available prediction or correction, despite being a perfect match.

This is a current issue with `kmc-model` as of `17.0.294-beta`.  I've determined that approximately 1 of every 36 words in the MTNT model (0.3.1) is affected by this issue, with other currently-released models similarly affected, sometimes at a rate of over 1 in every 18 words.  (MTNT used to have a similar error rate, but changes in MTNT 0.3.0 removed a lot of the entries that would have been affected by this.)

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #11073.

This fixes an issue where less-frequent words that match the start of higher-frequency words were inaccessible to the predictive-text engine and thus could never be suggested.  This effectively suggested to users that the word doesn't exist, which is certainly less than ideal when spelling assistance is desired.  Also, this is something that would certainly affect auto-correction efforts if and when the time comes.

A unit test is included.  The first commit (with the unit test) fails CI checks with 17.0.294-beta's current state and passes with the fix included (commit 2).

**Note**:  as this is a compiler bug, _all_ models will likely need recompilation so that they may receive the fix's effects.

## User Testing

**TEST_MTNT_THIN**:  Using the Keyman Developer test artifact, recompile the MTNT model and verify that #11073 is fixed.
1. Open Keyman Developer
2. Open the MTNT model's project (from within the keymanapp/lexical-models repo, path = release/nrc/nrc.en.mtnt)
3. Making no changes, build the model project.
4. Use Developer to test the model via the test-host page.
5. Then verify that `thin` shows up as a predictive-text suggestion when "thin" is typed.

### CODE REVIEW

File: developer/src/kmc-model/src/build-trie.ts [hunk 1]
```diff
@@ -374,6 +374,11 @@ namespace Trie {
    */
   function addItemToInternalNode(node: InternalNode, item: Entry, index: number) {
     let char = item.key[index];
+    // If an internal node is the proper site for item, it belongs under the
+    // corresponding (sentinel, internal-use) child node signifying this.
+    if(char === null || char === undefined) {
+      char = INTERNAL_VALUE;
+    }
```
Comment (jahorton) [inline developer/src/kmc-model/src/build-trie.ts:381]: Compare: 

https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/developer/src/kmc-model/src/build-trie.ts#L396

https://github.com/keymanapp/keyman/blob/acc38fea588986c70036571acdf2f334f83bf7b2/developer/src/kmc-model/src/build-trie.ts#L407-L419

**Comment by darcywong00** [comment]
> **Note**: as this is a compiler bug, all models will likely need recompilation so that they may receive the fix's effects.

Do we want to tack that onto keymanapp/lexical-models#227 (and bump the kps versions)?

**Comment by jahorton** [comment]
> > **Note**: as this is a compiler bug, all models will likely need recompilation so that they may receive the fix's effects.
> 
> Do we want to tack that onto [keymanapp/lexical-models#227](https://github.com/keymanapp/lexical-models/pull/227) (and bump the kps versions)?

Sounds wise.

**Comment by mcdurdin** [comment]
> Do we want to tack that onto [keymanapp/lexical-models#227](https://github.com/keymanapp/lexical-models/pull/227) (and bump the kps versions)?

I think we should keep the restructuring separate from the bug fix. Good to do it after the staging merge, but let's not mix the two processes.

**Comment by mcdurdin** [comment]
> This fixes an issue where less-frequent words that match the start of higher-frequency words were inaccessible to the predictive-text engine and thus could never be suggested. This effectively suggested to users that the word doesn't exist, which is certainly less than ideal when spelling assistance is desired. Also, this is something that would certainly affect auto-correction efforts if and when the time comes.

great description thanks!

File: developer/src/kmc-model/test/test-compile-trie.ts [hunk 2]
```diff
@@ -160,5 +160,15 @@ describe('createTrieDataStructure()', function () {
...
+  });
+
+  it('does not create `null`/"undefined"-keyed children', function () {
+    const WORDLIST_FILENAME = makePathToFixture('example.qaa.wordlist-ordering', 'wordlist.tsv');
+    // check to ensure the Trie is fully well-formed.
+    let sourceCode = createTrieDataStructure([WORDLIST_FILENAME], (wf) => wf);
+    assert.notMatch(sourceCode, /"undefined"/);
+    assert.notMatch(sourceCode, /null/);
+
+    console.log(sourceCode);
```
Comment (mcdurdin) [inline developer/src/kmc-model/test/test-compile-trie.ts:172]: Can we leave this out of the test?

Comment (mcdurdin) [reply developer/src/kmc-model/test/test-compile-trie.ts:172]:
```suggestion
    //console.log(sourceCode);
```

File: developer/src/kmc-model/src/build-trie.ts [hunk 3]
```diff
@@ -374,6 +374,11 @@ namespace Trie {
    */
   function addItemToInternalNode(node: InternalNode, item: Entry, index: number) {
     let char = item.key[index];
+    // If an internal node is the proper site for item, it belongs under the
+    // corresponding (sentinel, internal-use) child node signifying this.
+    if(char === null || char === undefined) {
```
Comment (mcdurdin) [inline developer/src/kmc-model/src/build-trie.ts:379]: This is a good catch but I am wondering how these two situations can arise -- i.e. why we need to test for both `null` and `undefined`?

File: developer/src/kmc-model/test/test-compile-trie.ts [hunk 4]
```diff
@@ -160,5 +160,15 @@ describe('createTrieDataStructure()', function () {
...
     assert.match(uppercaseSourceCode, /"key":\s*"TURTLES"/);
     assert.notMatch(uppercaseSourceCode, /"key":\s*"turtles"/);
+  });
+
+  it('does not create `null`/"undefined"-keyed children', function () {
+    const WORDLIST_FILENAME = makePathToFixture('example.qaa.wordlist-ordering', 'wordlist.tsv');
+    // check to ensure the Trie is fully well-formed.
+    let sourceCode = createTrieDataStructure([WORDLIST_FILENAME], (wf) => wf);
+    assert.notMatch(sourceCode, /"undefined"/);
+    assert.notMatch(sourceCode, /null/);
```
Comment (mcdurdin) [inline developer/src/kmc-model/test/test-compile-trie.ts:170]: This adequately tests the effect of the bug but it doesn't really test the actual validity of the outcome very well. I wonder if there is a more rigorous test we could apply?

**Review by mcdurdin — APPROVED** [review]
LGTM. Some minor questions. Good work.

Let's plan to rebuild the lexical models after 17.0 goes to release (I'll open an issue on the repo)

File: developer/src/kmc-model/test/test-compile-trie.ts (hunk 4 above)
Comment (jahorton) [reply developer/src/kmc-model/test/test-compile-trie.ts:170]: The obvious "most rigorous" approach would be to spin up a full `TrieModel` - or maybe just its 'traversal' - and ensure that all wordlist entries are indeed accessible.  That'd add a test dependency on `@keymanapp/models-templates`, though.  I can do that if we're fine with making that link.

Comment (mcdurdin) [reply developer/src/kmc-model/test/test-compile-trie.ts:170]: I think that a full TrieModel sounds like a good idea, but perhaps it should go in a separate PR, particularly as it adds an extra dependency? I am comfortable with this one merging as-is.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_MTNT_THIN (PASSED):** Tested with the attached PR build (Keyman Developer 17.0.295-beta-test-11074) on Windows 11 OS (VM) and here is my observation: 1. Opened nrc.en.mtnt project in the Keyman developer. 2. Clicked the build tab. 3. Compiled the model. 4. Opened the test-host page. 5. Selected sil_euro latin keyboard. 6. Selected 'nrc_en_mtnt' from the Model list box. Selected a Mobile device. 7. Typed 'thin' and verified that the word 'think (?)' showed as a predictive-text suggestion banner.

**Comment by jahorton** [comment]
> # Test Results
> 
> * **TEST_MTNT_THIN (PASSED):** Tested with the attached PR build (Keyman Developer 17.0.295-beta-test-11074) on Windows 10 OS and here is my observation: 1. Opened nrc.en.mtnt project in the Keyman developer. 2. Clicked the build tab. 3. Compiled the model. 4. Opened the test-host page. 5. Selected sil_euro latin keyboard. 6. Selected 'nrc_en_mtnt' from the Model list box. Selected a Mobile device. 7. Typed 'thin' and verified that the word 'think (?)' showed as a predictive-text suggestion banner.
>        

So, that would be a FAIL.  "think" is _**not**_ "thin".

That said, downloading the artifact for myself and carrying out the steps:

image

So, I'll let it slide here.

File: developer/src/kmc-model/test/test-compile-trie.ts (hunk 4 above)
Comment (jahorton) [reply developer/src/kmc-model/test/test-compile-trie.ts:170]: Sounds like a plan:  see #11085.
